### ISSUE TITLE
set spellsuggest=best,10 slows down spellchecking by >=15sec

### ISSUE THREAD

**Issue by ghost** [vim/vim#4087]
When adding a cap to the listed suggestion with set spellsuggest+=10, then spell checking is slowed down considerably letting vim freeze for >=15 sec (on my machine).
During that freeze vim uses 100% CPU time, the time taken also depends on the word to be corrected with `z=` (examples below). Note that the freeze can be interrupted with CTRL-C.
Obviously, such slow down renders the spelling correction feature nearly useless.

Also, adding the number 10 to spellsuggest does something more than just setting the max number of suggestions, it seems that it sometimes forces to list exactly 10 suggestions.

```
vim -u NORC
ispellcapcheck<ESC>
:set spell
:set spelllang=en " make sure to have the default dictionary installed
:set spellsuggest? " should be "best"
z=  " instant suggestions (about ~60)
:set spellsuggest=best,10
z=  " takes ~15sec before displaying 10 suggestions
:set spellsuggest=fast,10
z=  " noticeable delay (<1sec) before displaying 5(!) suggestions
:set spellsuggest=double,10
z=  " take ~20sec before displaying 10 suggestions
:set spellsuggest=10
z=  " noticeable delay (<1sec) before displaying 10 suggestions
```

All vanilla `spellsuggest` methods (best, fast, double) are instant.
When using only a number (`set spellsuggest=10`) it is unclear which method was used.
The suggestions do not coincide with any of the suggestion lists of `fast`, `double`, `best`.

To see that `set spellsuggest+=10` sometimes forces 10 suggestions, we have to switch out the dictionary (as I couldn't find a word with <10 suggestions for the dictionary shipped with Vim).
I was able to reproduce it with the English dictionary shipped with Neovim-0.3.3 .

```
cp neovim-en.utf-8.spl ~/.vim/spell/en.utf-8.spl
vim -u NORC
ispellcapcheck<ESC>
:set spell
:set spelllang=en
:set spellsuggest? " should be "best"
z= " lists 4 suggestions namely "Spell cap check", "Spell paycheck", "Spell hatcheck", "Skullcap check"
:set spellsuggest=best,10
z= " lists 10 suggestions (first four coincide with the above four)
:set spellsuggest=fast,10
z= " lists 4 suggestions
:set spellsuggest=double,10
z= " lists 10 suggestions
```

The methods `best`, `fast`, `double` list only 4 suggestions. However, when adding 10 to `spellsuggest`, then `best` and `double` result in 10 suggestions while `fast` does not.

I have not tested other methods of `spellsuggest` like `expr:...` or `file:...`.

**Comment by jdhao** [comment]
Same issue here on Neovim. Adding number to `spellsuggest` option makes it super slow. I do not know the implementation. I thought that a limit on the number of suggestions to show should make spell suggestion faster, not slower...

**Comment by chrisbra** [comment]
fixed by 77a849c
